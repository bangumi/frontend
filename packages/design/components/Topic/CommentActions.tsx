import React from 'react';

import { Comment as CommentIcon, More } from '@bangumi/icons';

import Button from '../../components/Button';
import Popover from '../Popover';

export interface CommentActionsProps {
  id: number;
  onReply?: () => void;
  onDelete?: () => void;
  isAuthor?: boolean;
  editable?: boolean;
  showText?: boolean;
}

const CommentActions = ({
  id,
  onReply,
  onDelete,
  isAuthor = false,
  editable = true,
  showText = false,
}: CommentActionsProps) => {
  return (
    <div className='bgm-comment-actions'>
      <Button type='plain' size='small' onClick={onReply} title='回复'>
        <CommentIcon />
        {showText && '回复'}
      </Button>
      {/* TODO: 实现贴贴功能 */}
      <Popover
        content={
          <div className='bgm-comment-actions__popover'>
            {isAuthor && (
              <>
                {editable && (
                  <Button.Link type='text' size='small' to={`/group/reply/${id}/edit`}>
                    编辑
                  </Button.Link>
                )}
                <Button type='text' size='small' onClick={onDelete}>
                  删除
                </Button>
              </>
            )}
            {/* TODO: 实现绝交和报告疑虑功能 */}
            <Button type='text'>绝交</Button>
            <Button type='text'>报告疑虑</Button>
          </div>
        }
      >
        <Button type='plain' size='small' className='bgm-comment-actions__more' title='其他'>
          <More />
          {showText && '其他'}
        </Button>
      </Popover>
    </div>
  );
};

export default CommentActions;
