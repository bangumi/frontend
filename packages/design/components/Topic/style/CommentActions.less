@import '../../../theme/base';

.bgm-comment-actions {
  display: flex;
  gap: 12px;

  .bgm-button {
    height: 20px;
    font-weight: 400;

    &.bgm-comment-actions__more:hover {
      color: @gray-80;
    }
  }

  &__popover {
    border-radius: 17px;
    display: flex;
    flex-direction: column;
    gap: 10px;
    align-items: center;
    padding: 10px;

    .bgm-button {
      font-size: 14px;
      height: 22px;
      border-radius: 11px;
      width: 70px;
      display: flex;
      justify-content: center;
      align-items: center;
      color: @gray-80;
      cursor: pointer;

      &:hover {
        background-color: #f5f5f5;
      }
    }
  }
}
