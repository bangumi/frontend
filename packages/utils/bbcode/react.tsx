import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
