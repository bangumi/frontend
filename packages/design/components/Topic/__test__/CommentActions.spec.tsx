import { render as _render } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import type { CommentActionsProps } from '../CommentActions';
import CommentActions from '../CommentActions';

const render = (props: CommentActionsProps) =>
  _render(
    <BrowserRouter>
      <CommentActions {...props} />
    </BrowserRouter>,
  );

describe('Basic Usage', () => {
  it('should render', () => {
    expect(render({ id: 233 })).toMatchSnapshot();
  });
});

describe('With Text', () => {
  it('should render', () => {
    expect(render({ id: 233, showText: true })).toMatchSnapshot();
  });
});

describe('Is Author', () => {
  it('should render', () => {
    expect(render({ id: 233, isAuthor: true })).toMatchSnapshot();
  });
});

describe('Non-editable', () => {
  it('should render', () => {
    expect(render({ id: 233, isAuthor: true, editable: false })).toMatchSnapshot();
  });
});
