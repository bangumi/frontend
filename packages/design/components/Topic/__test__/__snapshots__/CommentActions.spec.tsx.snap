// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Basic Usage > should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bgm-comment-actions"
      >
        <button
          class="bgm-button bgm-button--plain bgm-button--size-small"
          title="回复"
        >
          <div />
        </button>
        <div
          class="bgm-popover"
        >
          <button
            class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
            title="其他"
          >
            <div />
          </button>
          <div
            class="bgm-popover__container"
          >
            <div
              class="bgm-popover__content"
            >
              <div
                class="bgm-comment-actions__popover"
              >
                <button
                  class="bgm-button bgm-button--text"
                >
                  绝交
                </button>
                <button
                  class="bgm-button bgm-button--text"
                >
                  报告疑虑
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bgm-comment-actions"
    >
      <button
        class="bgm-button bgm-button--plain bgm-button--size-small"
        title="回复"
      >
        <div />
      </button>
      <div
        class="bgm-popover"
      >
        <button
          class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
          title="其他"
        >
          <div />
        </button>
        <div
          class="bgm-popover__container"
        >
          <div
            class="bgm-popover__content"
          >
            <div
              class="bgm-comment-actions__popover"
            >
              <button
                class="bgm-button bgm-button--text"
              >
                绝交
              </button>
              <button
                class="bgm-button bgm-button--text"
              >
                报告疑虑
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Is Author > should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bgm-comment-actions"
      >
        <button
          class="bgm-button bgm-button--plain bgm-button--size-small"
          title="回复"
        >
          <div />
        </button>
        <div
          class="bgm-popover"
        >
          <button
            class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
            title="其他"
          >
            <div />
          </button>
          <div
            class="bgm-popover__container"
          >
            <div
              class="bgm-popover__content"
            >
              <div
                class="bgm-comment-actions__popover"
              >
                <a
                  class="bgm-button bgm-button--link bgm-button--text bgm-button--size-small"
                  href="/group/reply/233/edit"
                >
                  编辑
                </a>
                <button
                  class="bgm-button bgm-button--text bgm-button--size-small"
                >
                  删除
                </button>
                <button
                  class="bgm-button bgm-button--text"
                >
                  绝交
                </button>
                <button
                  class="bgm-button bgm-button--text"
                >
                  报告疑虑
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bgm-comment-actions"
    >
      <button
        class="bgm-button bgm-button--plain bgm-button--size-small"
        title="回复"
      >
        <div />
      </button>
      <div
        class="bgm-popover"
      >
        <button
          class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
          title="其他"
        >
          <div />
        </button>
        <div
          class="bgm-popover__container"
        >
          <div
            class="bgm-popover__content"
          >
            <div
              class="bgm-comment-actions__popover"
            >
              <a
                class="bgm-button bgm-button--link bgm-button--text bgm-button--size-small"
                href="/group/reply/233/edit"
              >
                编辑
              </a>
              <button
                class="bgm-button bgm-button--text bgm-button--size-small"
              >
                删除
              </button>
              <button
                class="bgm-button bgm-button--text"
              >
                绝交
              </button>
              <button
                class="bgm-button bgm-button--text"
              >
                报告疑虑
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Non-editable > should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bgm-comment-actions"
      >
        <button
          class="bgm-button bgm-button--plain bgm-button--size-small"
          title="回复"
        >
          <div />
        </button>
        <div
          class="bgm-popover"
        >
          <button
            class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
            title="其他"
          >
            <div />
          </button>
          <div
            class="bgm-popover__container"
          >
            <div
              class="bgm-popover__content"
            >
              <div
                class="bgm-comment-actions__popover"
              >
                <button
                  class="bgm-button bgm-button--text bgm-button--size-small"
                >
                  删除
                </button>
                <button
                  class="bgm-button bgm-button--text"
                >
                  绝交
                </button>
                <button
                  class="bgm-button bgm-button--text"
                >
                  报告疑虑
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bgm-comment-actions"
    >
      <button
        class="bgm-button bgm-button--plain bgm-button--size-small"
        title="回复"
      >
        <div />
      </button>
      <div
        class="bgm-popover"
      >
        <button
          class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
          title="其他"
        >
          <div />
        </button>
        <div
          class="bgm-popover__container"
        >
          <div
            class="bgm-popover__content"
          >
            <div
              class="bgm-comment-actions__popover"
            >
              <button
                class="bgm-button bgm-button--text bgm-button--size-small"
              >
                删除
              </button>
              <button
                class="bgm-button bgm-button--text"
              >
                绝交
              </button>
              <button
                class="bgm-button bgm-button--text"
              >
                报告疑虑
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`With Text > should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bgm-comment-actions"
      >
        <button
          class="bgm-button bgm-button--plain bgm-button--size-small"
          title="回复"
        >
          <div />
          回复
        </button>
        <div
          class="bgm-popover"
        >
          <button
            class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
            title="其他"
          >
            <div />
            其他
          </button>
          <div
            class="bgm-popover__container"
          >
            <div
              class="bgm-popover__content"
            >
              <div
                class="bgm-comment-actions__popover"
              >
                <button
                  class="bgm-button bgm-button--text"
                >
                  绝交
                </button>
                <button
                  class="bgm-button bgm-button--text"
                >
                  报告疑虑
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bgm-comment-actions"
    >
      <button
        class="bgm-button bgm-button--plain bgm-button--size-small"
        title="回复"
      >
        <div />
        回复
      </button>
      <div
        class="bgm-popover"
      >
        <button
          class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
          title="其他"
        >
          <div />
          其他
        </button>
        <div
          class="bgm-popover__container"
        >
          <div
            class="bgm-popover__content"
          >
            <div
              class="bgm-comment-actions__popover"
            >
              <button
                class="bgm-button bgm-button--text"
              >
                绝交
              </button>
              <button
                class="bgm-button bgm-button--text"
              >
                报告疑虑
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
